### ISSUE TITLE
Improvements to session handling

### ISSUE THREAD

**Issue by EiNSTeiN-** [Shopify/shopify_app#350]
I'm opening this issue to offer visibility and track some upcoming changes to session handling in this gem.
- [ ] Better multi-tenant support. Since shopify is a multi-tenant platform, apps must support merchants who are logged into their app from multiple stores at the same time, and this app does not make that easy. I'd like to make this work better out-of-the-box by scoping routes with a unique prefix (the shop id). We can also scope the session cookie's `path` to this unique prefix, removing the need to track multiple users in the same `session` object.
- [ ] Respect user permissions by default. We're introducing a way for users to login as themselves into apps. This should become the default for Embedded Apps, because these are meant to integrate seamlessly into the admin area, so respecting the user's permission is an obvious must-have.
- [ ] Keep track of extra information returned by Shopify, such as scope & associated user. Currently this app has no way to keep track of this, it only keeps the access token. Keeping track of the scope would facilitate the workflow when updating the application's requested scope. We should force a new oauth autorization when the needed access scope change.
- [ ] Longer session cookie expiry. The main reason to have short session expiry is to make sure the user still has access to the store, since the permanent access tokens never expire. However there's no reason we need to do this by loosing the session, we can easily keep the session forever and simply force a new login after a week or so (or until redis pressure evicts the session). 
- [ ] The new per-user access tokens expire within a rather short timeframe (24 hours) so in addition to the point above, we need to make sure to handle the case where the user was authorized correctly, but the access token becomes invalid. We know the expiry time of the token, but it may expire quicker than the stated expiry time due to various user actions within the shopify admin.
- [ ] Handle users missing the correct permissions (403 Forbidden response) in a sane way. This is the response Shopify will send when a per-user access token is associated with a user that does not have permission to access some data (for example, reading orders on behalf of a user that lacks the Orders permission).

**Comment by MatiasFMolinari** [comment]
Heya folks,

I'm building an app that requires users to sign up individually and have various roles with permissions. So far, I'm creating a `User` for each logged in Shopify user with the Shopify id received from  `session[:shopify_user]` as the unique identifier.

My `omniauth.rb` looks like this: https://gist.github.com/MatiasFMolinari/cf34e4fc18b46848e2de1b345d76e478

I believe this setup is only supporting online access, not offline. Therefore, **I assume the GDPR webhooks will not run when requested.** 

**Is this assumption correct or do those get authenticated differently?** If they do, then that's great! If not, I have one more question:

**How should `omniauth.rb` look like if I want to support both online and offline access?** Here's what I have so far, but I'm still not sure if it'll work as intended: https://gist.github.com/MatiasFMolinari/75b5efeef66a7d59e62a2bb42c5d8045

I assume that I'd need to call the sessions differently, changing `['shopify.omniauth_params']` to `['shopify.omniauth_user_params']`, or something like that?

Thanks in advance for the help!

**Comment by fredngo** [comment]
> The new per-user access tokens expire within a rather short timeframe (24 hours) so in addition to the point above, we need to make sure to handle the case where the user was authorized correctly, but the access token becomes invalid. We know the expiry time of the token, but it may expire quicker than the stated expiry time due to various user actions within the shopify admin.

I'm wondering why this issue was closed? I am running into this 24 hour token expiry problem.

I have found that if the admin user does not login to my app (via the Shopify dashboard) at least once every 24 hours, background API requests (to fetch product data, etc) start failing due to token expiry (i.e. Unauthorized). 

This is not ideal as it would not be normal for a human user to have to login to the app so often -- Is there anyway to deal with this? Is there a method to refresh the token? I'm poring over the source code for the gem and can't seem to find anything to help here -- I could be wrong, I'm quite a newbie when it comes to Shopify.
